ambry-router/src/main/java/com.github.ambry.router/CloudRouterFactory.java

```diff
@@ -0,0 +1,163 @@
+    defaultPartitionClass = clusterMapConfig.clusterMapDefaultPartitionClass;
+
+    // TODO: need to shut pool down on exit, may need to move to RestServer
+    // Or pass it to router
+    requestHandlerPool = getRequestHandlerPool(verifiableProperties, clusterMap, notificationSystem);
+    networkClientFactory = new LocalNetworkClientFactory((LocalRequestResponseChannel) requestHandlerPool.getChannel(),
+        new NetworkConfig(verifiableProperties), new NetworkMetrics(registry), time);
+    logger.info("Instantiated CloudRouterFactory");
+  }
+
+  public static RequestHandlerPool getRequestHandlerPool(VerifiableProperties verifiableProperties,
+      ClusterMap clusterMap, NotificationSystem notificationSystem) throws Exception {
+    ClusterMapConfig clusterMapConfig = new ClusterMapConfig(verifiableProperties);
+    CloudConfig cloudConfig = new CloudConfig(verifiableProperties);
+    // TODO: move properties into maybe routerConfig, frontendConfig
+    ServerConfig serverConfig = new ServerConfig(verifiableProperties);
+    StoreConfig storeConfig = new StoreConfig(verifiableProperties);
+    MetricRegistry registry = clusterMap.getMetricRegistry();
+
+    DataNodeId nodeId = new CloudDataNode(cloudConfig, clusterMapConfig);
+    CloudDestinationFactory cloudDestinationFactory =
+        Utils.getObj(cloudConfig.cloudDestinationFactoryClass, verifiableProperties, registry);
+    CloudDestination cloudDestination = cloudDestinationFactory.getCloudDestination();
+    VcrMetrics vcrMetrics = new VcrMetrics(registry);
+    StoreManager cloudStorageManager =
+        new CloudStorageManager(verifiableProperties, vcrMetrics, cloudDestination, clusterMap);
+    LocalRequestResponseChannel channel = new LocalRequestResponseChannel();
+    ServerMetrics serverMetrics = new ServerMetrics(registry, AmbryRequests.class);
+    StoreKeyFactory storeKeyFactory = Utils.getObj(storeConfig.storeKeyFactory, clusterMap);
+    StoreKeyConverterFactory storeKeyConverterFactory =
+        Utils.getObj(serverConfig.serverStoreKeyConverterFactory, verifiableProperties, registry);
+    AmbryRequests requests =
+        new AmbryRequests(cloudStorageManager, channel, clusterMap, nodeId, registry, serverMetrics, null,
+            notificationSystem, null, storeKeyFactory, serverConfig.serverEnableStoreDataPrefetch,
+            storeKeyConverterFactory);
+    return new RequestHandlerPool(serverConfig.serverRequestHandlerNumOfThreads, channel, requests);
+  }
+
+  /**
+   * Construct and return a {@link NonBlockingRouter} that works with cloud storage.
+   * @return a {@link NonBlockingRouter}
+   */
+  @Override
+  public Router getRouter() {
+    try {
+      NonBlockingRouter router = new NonBlockingRouter(routerConfig, routerMetrics, networkClientFactory, notificationSystem, clusterMap,
+          kms, cryptoService, cryptoJobHandler, accountService, time, defaultPartitionClass);
+      router.setRequestHandlerPool(requestHandlerPool);
```

Comment:
I don't see requestHandlerPool is called in router, will it be used in future PR?

Reply:
The NonBlockingRouter closes it on shutdown.  It may not the the ideal place to do it but I couldn't think of a better place to put it.  The factory itself has no shutdown hook.

Reply:
Cant we just pass requestHandlerPool in router's constructor?

Reply:
Done.